File: gopher-mcp/sdk/typescript/mcp-example/src/gopher-transport.ts
import { Transport, TransportSendOptions } from "@modelcontextprotocol/sdk/shared/transport";
import { JSONRPCMessage, MessageExtraInfo } from "@modelcontextprotocol/sdk/types";
import {
  JSONRPCMessage as FilterJSONRPCMessage,
  FilterManager,
  FilterManagerConfig,
} from "../../src/mcp-filter-manager";

// Type adapter to convert between MCP SDK and FilterManager JSONRPCMessage types
type MCPJSONRPCMessage = JSONRPCMessage;
type AdapterJSONRPCMessage = FilterJSONRPCMessage;

function adaptToFilterMessage(mcpMessage: MCPJSONRPCMessage): AdapterJSONRPCMessage {
  return mcpMessage as AdapterJSONRPCMessage;
}

function adaptToMCPMessage(filterMessage: AdapterJSONRPCMessage): MCPJSONRPCMessage {
  return filterMessage as MCPJSONRPCMessage;
}

export interface GopherTransportConfig {
  // Filter configuration for this transport
  filters?: FilterManagerConfig;

  // Transport-specific configuration
  name?: string;
  version?: string;

  // Connection configuration
  host?: string;
  port?: number;
  protocol?: "tcp" | "udp" | "stdio";

  // Timeout configuration
  connectTimeout?: number;
  sendTimeout?: number;
  receiveTimeout?: number;
}

export class GopherTransport implements Transport {
  private filterManager: FilterManager;
  private config: GopherTransportConfig;
  private isConnected: boolean = false;
  private isDestroyed: boolean = false;
  private server: any = null; // For server mode
  private client: any = null; // For client mode
  private connections: Map<string, any> = new Map(); // Active connections
  private messageBuffers: Map<string, string> = new Map(); // Message buffers for each connection

  // Transport event handlers
  onclose?: () => void;
  onerror?: (error: Error) => void;
  onmessage?: (message: JSONRPCMessage, extra?: MessageExtraInfo) => void;
  sessionId?: string;
  setProtocolVersion?: (version: string) => void;

  constructor(config: GopherTransportConfig = {}) {
    this.config = {
      name: "gopher-transport",
      version: "1.0.0",
      protocol: "stdio",
      connectTimeout: 30000, // 30 seconds
      sendTimeout: 10000, // 10 seconds
      receiveTimeout: 30000, // 30 seconds
      ...config,
    };

    // Initialize FilterManager with transport-specific configuration
    const filterConfig: FilterManagerConfig = {
      // Default security configuration
      security: {
        authentication: {
          method: "jwt",
          secret: "gopher-transport-secret",
        },
        authorization: {
          enabled: true,
          policy: "allow",
          rules: [
            {
              resource: "*",
              action: "read",
            },
          ],
        },
      },

      // Default observability configuration
      observability: {
        accessLog: {
          enabled: true,
          format: "json",
          fields: ["timestamp", "method", "sessionId", "duration"],
          output: "console",
        },
        metrics: {
          enabled: true,
          labels: {
            transport: this.config.name || "gopher-transport",
            version: this.config.version || "1.0.0",
          },
        },
        tracing: {
          enabled: true,
          serviceName: `gopher-transport-${this.config.name || "default"}`,
          samplingRate: 0.1, // 10% sampling
        },
      },

      // Default traffic management
      trafficManagement: {
        rateLimit: {
          enabled: true,
          requestsPerMinute: 1000,
          burstSize: 50,
          keyExtractor: "custom",
        },
        circuitBreaker: {
          enabled: true,
          failureThreshold: 5,
          timeout: 30000,
          resetTimeout: 60000,
        },
        retry: {
          enabled: true,
          maxAttempts: 3,
          backoffStrategy: "exponential",
          baseDelay: 1000,
          maxDelay: 5000,
        },
      },

      // Default error handling
      errorHandling: {
        stopOnError: false,
        retryAttempts: 2,
        fallbackBehavior: "passthrough",
      },

      // Merge with user-provided configuration
      ...this.config.filters,
    };

    this.filterManager = new FilterManager(filterConfig);
    console.log(`🔧 GopherTransport initialized with FilterManager (${this.config.name})`);
  }

  async start(): Promise<void> {
    if (this.isDestroyed) {
      throw new Error("GopherTransport has been destroyed");
    }

    if (this.isConnected) {
      console.warn("GopherTransport is already connected");
      return;
    }

    try {
      console.log(`🚀 Starting GopherTransport (${this.config.protocol})`);

      // Generate session ID
      this.sessionId = this.generateSessionId();
      console.log(`📋 Session ID: ${this.sessionId}`);

      // Set protocol version
      if (this.setProtocolVersion) {
        this.setProtocolVersion("2024-11-05");
      }

      // Start transport based on protocol
      await this.startTransport();

      this.isConnected = true;
      console.log("✅ GopherTransport started successfully");
    } catch (error) {
      this.isConnected = false;
      const errorMsg = `Failed to start GopherTransport: ${error}`;
      console.error("❌", errorMsg);

      if (this.onerror) {
        this.onerror(new Error(errorMsg));
      }
      throw error;
    }
  }

  async send(message: JSONRPCMessage, _options?: TransportSendOptions): Promise<void> {
    if (this.isDestroyed) {
      throw new Error("GopherTransport has been destroyed");
    }

    if (!this.isConnected) {
      throw new Error("GopherTransport is not connected");
    }

    try {
      const messageInfo =
        "method" in message
          ? `${message.method} (id: ${"id" in message ? message.id : "N/A"})`
          : "notification";
      console.log(`📤 Sending message: ${messageInfo}`);

      // Process message through FilterManager
      const filterMessage = adaptToFilterMessage(message);
      const processedFilterMessage = await this.filterManager.process(filterMessage);
      const processedMessage = adaptToMCPMessage(processedFilterMessage);

      const processedInfo =
        "method" in processedMessage
          ? `${processedMessage.method} (id: ${
              "id" in processedMessage ? processedMessage.id : "N/A"
            })`
          : "notification";
      console.log(`✅ Message processed through filters: ${processedInfo}`);

      // FilterManager processing complete - message ready for actual transport
      console.log(`✅ Message ready for transport: ${JSON.stringify(processedMessage, null, 2)}`);

      // Send the processed message through the actual transport
      await this.sendThroughTransport(processedMessage);
    } catch (error) {
      const errorMsg = `Failed to send message: ${error}`;
      console.error("❌", errorMsg);

      if (this.onerror) {
        this.onerror(new Error(errorMsg));
      }
      throw error;
    }
  }

  async close(): Promise<void> {
    if (this.isDestroyed) {
      console.warn("GopherTransport is already destroyed");
      return;
    }

    try {
      console.log("🔌 Closing GopherTransport connection");

      this.isConnected = false;

      // Close transport connections
      if (this.client) {
        this.client.destroy();
        this.client = null;
      }

      if (this.server) {
        this.server.close();
        this.server = null;
      }

      // Close all active connections
      for (const [_connectionId, connection] of this.connections) {
        if (connection.destroy) {
          connection.destroy();
        }
      }
      this.connections.clear();
      this.messageBuffers.clear();

      // Clean up FilterManager resources
      this.filterManager.destroy();

      this.isDestroyed = true;
      console.log("✅ GopherTransport closed successfully");

      // Notify close event
      if (this.onclose) {
        this.onclose();
      }
    } catch (error) {
      console.error("❌ Error closing GopherTransport:", error);
      this.isDestroyed = true; // Mark as destroyed even if cleanup failed
      throw error;
    }
  }

  /**
   * Process a received message through FilterManager
   * In a real implementation, this would be called when receiving data from the network
   */
  async processReceivedMessage(message: JSONRPCMessage, extra?: MessageExtraInfo): Promise<void> {
    if (!this.isConnected || this.isDestroyed) {
      return;
    }

    try {
      const messageInfo =
        "method" in message
          ? `${message.method} (id: ${"id" in message ? message.id : "N/A"})`
          : "notification";
      console.log(`📥 Processing received message: ${messageInfo}`);

      // Process response through FilterManager
      const filterMessage = adaptToFilterMessage(message);
      const processedFilterMessage = await this.filterManager.processResponse(filterMessage);
      const processedMessage = adaptToMCPMessage(processedFilterMessage);

      const processedInfo =
        "method" in processedMessage
          ? `${processedMessage.method} (id: ${
              "id" in processedMessage ? processedMessage.id : "N/A"
            })`
          : "notification";
      console.log(`✅ Response processed through filters: ${processedInfo}`);

      // Notify message event
      if (this.onmessage) {
        this.onmessage(processedMessage, extra);
      }
    } catch (error) {
      console.error("❌ Error processing received message:", error);

      if (this.onerror) {
        this.onerror(new Error(`Failed to process received message: ${error}`));
      }
    }
  }

  /**
   * Generate a unique session ID
   */
  private generateSessionId(): string {
    const timestamp = Date.now().toString(36);
    const random = Math.random().toString(36).substring(2);
    return `gopher-${timestamp}-${random}`;
  }

  /**
   * Get transport statistics
   */
  getStats() {
    return {
      isConnected: this.isConnected,
      isDestroyed: this.isDestroyed,
      sessionId: this.sessionId,
      config: this.config,
      connections: this.connections.size,
    };
  }

  /**
   * Start the actual transport layer based on protocol
   */
  private async startTransport(): Promise<void> {
    switch (this.config.protocol) {
      case "stdio":
        await this.startStdioTransport();
        break;
      case "tcp":
        await this.startTcpTransport();
        break;
      case "udp":
        await this.startUdpTransport();
        break;
      default:
        throw new Error(`Unsupported protocol: ${this.config.protocol}`);
    }
  }

  /**
   * Start stdio transport (for direct communication)
   */
  private async startStdioTransport(): Promise<void> {
    console.log("📡 Starting stdio transport");

    // For stdio, we'll use process.stdin/stdout for real communication
    process.stdin.setEncoding("utf8");

    let buffer = "";

    process.stdin.on("data", async (chunk: string) => {
      buffer += chunk;

      // Process complete messages (delimited by newlines)
      const lines = buffer.split("\n");
      buffer = lines.pop() || ""; // Keep incomplete line in buffer

      for (const line of lines) {
        if (line.trim()) {
          try {
            const message = JSON.parse(line.trim());
            await this.processReceivedMessage(message);
          } catch (error) {
            console.error("❌ Error processing stdio message:", error);
          }
        }
      }
    });

    process.stdin.on("end", () => {
      console.log("📡 Stdio input ended");
      if (this.onclose) {
        this.onclose();
      }
    });

    console.log("✅ Stdio transport ready for input");
  }

  /**
   * Start TCP transport
   */
  private async startTcpTransport(): Promise<void> {
    console.log(`📡 Starting TCP transport on ${this.config.host}:${this.config.port}`);

    // Import net module for TCP
    const net = await import("net");

    if (this.config.host && this.config.port) {
      // Client mode - connect to server
      this.client = new net.Socket();

      this.client.on("connect", () => {
        console.log("🔗 Connected to TCP server");
      });

      this.client.on("data", async (data: Buffer) => {
        await this.processTcpData(data, "client");
      });

      this.client.on("close", () => {
        console.log("🔌 TCP connection closed");
        if (this.onclose) {
          this.onclose();
        }
      });

      this.client.on("error", (error: Error) => {
        console.error("❌ TCP client error:", error);
        if (this.onerror) {
          this.onerror(error);
        }
      });

      // Connect to server
      await new Promise<void>((resolve, reject) => {
        this.client.connect(this.config.port!, this.config.host!, (error?: Error) => {
          if (error) {
            reject(error);
          } else {
            resolve();
          }
        });
      });
    } else {
      // Server mode - listen for connections
      this.server = net.createServer(socket => {
        const connectionId = `${socket.remoteAddress}:${socket.remotePort}`;
        console.log(`🔗 New TCP connection: ${connectionId}`);

        this.connections.set(connectionId, socket);

        socket.on("data", async (data: Buffer) => {
          await this.processTcpData(data, connectionId);
        });

        socket.on("close", () => {
          console.log(`🔌 TCP connection closed: ${connectionId}`);
          this.connections.delete(connectionId);
          this.messageBuffers.delete(connectionId);
        });

        socket.on("error", (error: Error) => {
          console.error(`❌ TCP connection error (${connectionId}):`, error);
          this.connections.delete(connectionId);
          this.messageBuffers.delete(connectionId);
        });
      });

      const port = this.config.port || 8080;
      await new Promise<void>((resolve, reject) => {
        this.server.listen(port, (error?: Error) => {
          if (error) {
            reject(error);
          } else {
            console.log(`🚀 TCP server listening on port ${port}`);
            resolve();
          }
        });
      });
    }
  }

  /**
   * Start UDP transport
   */
  private async startUdpTransport(): Promise<void> {
    console.log(`📡 Starting UDP transport on ${this.config.host}:${this.config.port}`);

    // Import dgram module for UDP
    const dgram = await import("dgram");

    if (this.config.host && this.config.port) {
      // Client mode
      this.client = dgram.createSocket("udp4");

      this.client.on("message", async (msg: Buffer, _rinfo: any) => {
        try {
          const message = JSON.parse(msg.toString());
          await this.processReceivedMessage(message);
        } catch (error) {
          console.error("❌ Error processing received UDP data:", error);
        }
      });

      this.client.on("error", (error: Error) => {
        console.error("❌ UDP client error:", error);
        if (this.onerror) {
          this.onerror(error);
        }
      });
    } else {
      // Server mode
      this.server = dgram.createSocket("udp4");

      this.server.on("message", async (msg: Buffer, rinfo: any) => {
        try {
          const message = JSON.parse(msg.toString());
          const connectionId = `${rinfo.address}:${rinfo.port}`;
          this.connections.set(connectionId, rinfo);
          await this.processReceivedMessage(message);
        } catch (error) {
          console.error("❌ Error processing received UDP data:", error);
        }
      });

      this.server.on("error", (error: Error) => {
        console.error("❌ UDP server error:", error);
        if (this.onerror) {
          this.onerror(error);
        }
      });

      const port = this.config.port || 8080;
      this.server.bind(port, () => {
        console.log(`🚀 UDP server listening on port ${port}`);
      });
    }
  }

  /**
   * Process TCP data with proper message delimiters
   */
  private async processTcpData(data: Buffer, connectionId: string): Promise<void> {
    // Get or create buffer for this connection
    let buffer = this.messageBuffers.get(connectionId) || "";
    buffer += data.toString();

    // Process complete messages (delimited by newlines)
    const lines = buffer.split("\n");
    buffer = lines.pop() || ""; // Keep incomplete line in buffer
    this.messageBuffers.set(connectionId, buffer);

    for (const line of lines) {
      if (line.trim()) {
        try {
          const message = JSON.parse(line.trim());
          await this.processReceivedMessage(message);
        } catch (error) {
          console.error(`❌ Error processing TCP message from ${connectionId}:`, error);
        }
      }
    }
  }

  /**
   * Send message through the actual transport
   */
  private async sendThroughTransport(message: JSONRPCMessage): Promise<void> {
    const messageData = JSON.stringify(message) + "\n"; // Add newline delimiter

    switch (this.config.protocol) {
      case "stdio":
        // For stdio, write to stdout
        process.stdout.write(messageData);
        break;

      case "tcp":
        if (this.client) {
          // Client mode - send to server
          this.client.write(messageData);
        } else if (this.server) {
          // Server mode - broadcast to all connections
          for (const [_connectionId, socket] of this.connections) {
            socket.write(messageData);
          }
        }
        break;

      case "udp":
        if (this.client) {
          // Client mode - send to server
          this.client.send(messageData, this.config.port!, this.config.host!);
        } else if (this.server) {
          // Server mode - broadcast to all connections
          for (const [_connectionId, rinfo] of this.connections) {
            this.server.send(messageData, rinfo.port, rinfo.address);
          }
        }
        break;

      default:
        throw new Error(`Unsupported protocol for sending: ${this.config.protocol}`);
    }
  }
}

File: gopher-mcp/sdk/ruby/lib/mcp_filter_sdk/gopher_transport.rb
# frozen_string_literal: true

require 'socket'
require 'json'
require 'securerandom'
require 'mcp_filter_sdk/types/index'

module McpFilterSdk
  class GopherTransport
    include Types

    attr_reader :config, :connections, :filters, :session_id, :is_connected, :is_destroyed

    def initialize(config)
      @config = config.is_a?(Hash) ? TransportConfig.new(config) : config
      @connections = {}
      @filters = []
      @session_id = "gopher-#{Time.now.to_i}-#{SecureRandom.random_number(2**64)}"
      @is_connected = false
      @is_destroyed = false
      @tcp_listener = nil
      @tcp_server_thread = nil
    end

    def start
      puts "🚀 Starting GopherTransport (#{@config.protocol.to_s.capitalize})"
      puts "📋 Session ID: #{@session_id}"

      case @config.protocol
      when :tcp
        start_tcp_transport
      when :udp
        start_udp_transport
      when :stdio
        start_stdio_transport
      else
        raise TransportError.new(-1, "Unsupported protocol: #{@config.protocol}")
      end

      @is_connected = true
      puts '✅ GopherTransport started successfully'
    end

    def stop
      puts '🔌 Closing GopherTransport connection'

      @tcp_server_thread&.kill
      @tcp_listener&.close
      @connections.each_value(&:close)
      @connections.clear

      @is_connected = false
      @is_destroyed = true

      puts '✅ GopherTransport closed successfully'
    end

    def send_message(message)
      # Check if transport is connected
      unless @is_connected
        puts '❌ Cannot send message: transport not connected'
        return false
      end

      # Handle both string and hash messages
      message_hash = if message.is_a?(String)
                       { method: 'test', id: 1, params: { data: message } }
                     else
                       message
                     end

      puts "📤 Sending message: #{message_hash[:method]} (id: #{message_hash[:id]})"

      # Process through filters
      processed_message = process_through_filters(message_hash)
      puts "✅ Message processed through filters: #{processed_message[:method]} (id: #{processed_message[:id]})"

      # Convert to JSON
      json_message = JSON.generate(processed_message)
      puts "✅ Message ready for transport: #{json_message}"

      # Send based on protocol
      result = case @config.protocol
               when :tcp
                 send_tcp_message(json_message)
               when :udp
                 send_udp_message(json_message)
               when :stdio
                 send_stdio_message(json_message)
               end

      puts '✅ Sent message through transport'
      result
    end

    def add_filter(filter)
      @filters << filter
      puts "✅ Added filter: #{filter.name}"
    end

    def get_stats
      {
        config: @config.to_h,
        connections: @connections.size,
        filters: @filters.size,
        is_connected: @is_connected,
        is_destroyed: @is_destroyed,
        session_id: @session_id
      }
    end

    private

    def start_tcp_transport
      host = @config.host || '127.0.0.1'
      port = @config.port || 8080

      puts "📡 Starting TCP transport on #{host}:#{port}"

      if @config.host.nil?
        # Server mode - listen for connections
        @tcp_listener = TCPServer.new(host, port)
        puts "🚀 TCP server listening on port #{port}"
        puts '🔄 Starting TCP server loop to accept connections...'

        @tcp_server_thread = Thread.new do
          loop do
            client = @tcp_listener.accept
            client_info = client.peeraddr
            connection_id = "#{client_info[3]}:#{client_info[1]}"

            puts "🔗 New connection from #{connection_id}"
            @connections[connection_id] = client
            puts "✅ Connection stored, total connections: #{@connections.size}"

            # Handle client in a separate thread
            Thread.new(client, connection_id) do |client_socket, conn_id|
              handle_tcp_client(client_socket, conn_id)
            end
          rescue StandardError => e
            puts "❌ Error accepting connection: #{e.message}"
          end
        end
      else
        # Client mode - connect to server
        client = TCPSocket.new(host, port)
        connection_id = "#{host}:#{port}"
        @connections[connection_id] = client
        puts '🔗 Connected to TCP server'
      end
    end

    def start_udp_transport
      host = @config.host || '127.0.0.1'
      port = @config.port || 8080

      puts "📡 Starting UDP transport on #{host}:#{port}"

      @udp_socket = UDPSocket.new
      if @config.host.nil?
        # Server mode
        @udp_socket.bind(host, port)
        puts "🚀 UDP server listening on port #{port}"
      else
        # Client mode
        @udp_socket.connect(host, port)
        puts '🔗 Connected to UDP server'
      end
    end

    def start_stdio_transport
      puts '📡 Starting stdio transport'
      puts '✅ Stdio transport ready for input'
    end

    def send_tcp_message(message)
      if @config.host.nil?
        # Server mode - broadcast to all clients
        @connections.each do |connection_id, client|
          client.puts(message)
          puts "📤 Sent to client #{connection_id}"
        rescue StandardError => e
          puts "❌ Error sending to client #{connection_id}: #{e.message}"
          @connections.delete(connection_id)
        end
      else
        # Client mode - send to server
        client = @connections.values.first
        client&.puts(message)
      end
    end

    def send_udp_message(message)
      @udp_socket&.send(message, 0)
    end

    def send_stdio_message(message)
      puts message
      true
    end

    def handle_tcp_client(client_socket, connection_id)
      while (line = client_socket.gets)
        puts "📥 Received from #{connection_id}: #{line.chomp}"
        # Process received message
      end
    rescue StandardError => e
      puts "❌ Error handling client #{connection_id}: #{e.message}"
    ensure
      client_socket.close
      @connections.delete(connection_id)
      puts "🔌 Client #{connection_id} disconnected"
    end

    def process_through_filters(message)
      processed_message = message.dup

      @filters.each do |filter|
        if filter.respond_to?(:process_data)
          # For CApiFilter, we need to extract the data from the hash
          if processed_message.is_a?(Hash) && processed_message[:params] && processed_message[:params][:data]
            # Extract data from hash message and process it
            data = processed_message[:params][:data]
            processed_data = filter.process_data(data)
            processed_message[:params][:data] = processed_data if processed_data
          else
            # Handle string message
            processed_data = filter.process_data(processed_message)
            processed_message = processed_data if processed_data
          end
        elsif filter.respond_to?(:callbacks) && filter.callbacks[:on_data]
          # Handle callback-based filters
          if processed_message.is_a?(Hash) && processed_message[:params] && processed_message[:params][:data]
            # Extract data from hash message
            data = processed_message[:params][:data]
            processed_data = filter.callbacks[:on_data].call(data)
            processed_message[:params][:data] = processed_data
          else
            # Handle string message
            processed_data = filter.callbacks[:on_data].call(processed_message)
            processed_message = processed_data
          end
        end
      rescue StandardError => e
        puts "❌ Error in filter #{filter.name}: #{e.message}"
      end

      processed_message
    end
  end
end
